### PR TITLE
Add route based ACL support

### PR DESCRIPTION
**Pull request by raymondfeng**
/to @ritch 
/cc @superkhau 

This PR adds route based ACL as an express middleware.

### CODE REVIEW

**Comment by raymondfeng** [comment]
Connect to strongloop/strong-gateway#38

File: server/middleware/acl.js [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+      }
+      if (route.regexp.test(url)) {
+        matchedScopes.push(route.scope);
+      }
+    }
+  }
+  return matchedScopes;
+}
+
+/**
```
Comment (ritch) [inline server/middleware/acl.js]: Mark this as private. Shouldn't include this in the docs.

Comment (ritch) [inline server/middleware/acl.js]: > Calculate the ACL score based its specifics of principalType and permission

Should be 

> Calculate the ACL score based **on** its specifics of principalType and permission

File: test/route-acl.test.js [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  var handler;
+
+  before(function(done) {
+    handler = routeAcl(options);
+    var ds = loopback.createDataSource({
+      connector: 'memory',
+      name: 'db',
+      defaultForType: 'db'
+    });
+    loopback.Application.attachTo(ds);
```
Comment (ritch) [inline test/route-acl.test.js]: We shouldn't use this soon to be deprecated version of core models... Use the new `registry` approach.

Comment (ritch) [inline test/route-acl.test.js]: A good rule of thumb is to not use `loopback.Application` directly in your test app.

File: test/route-acl.test.js [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+      'methods': 'GET',
+      'path': '/api/catalog'
+    }],
+    order: [{
+      methods: 'ALL',
+      path: '/api/orders'
+    }],
+    cancelOrder: [{
+      methods: 'delete',
+      'path': '/api/orders/:id'
```
Comment (ritch) [inline test/route-acl.test.js]: inconsistent quoting???

**Comment by ritch** [comment]
@raymondfeng 

Can you post an example and explain how this would work along side the existing ACL behavior?

Also I do not think we should add this into core. It looks like it might be convenient now, but its a brand new feature that I imagine will change quite a bit. If we land it in core now, we cannot make those changes until loopback 3. That approach (landing it in core now) will make this very difficult to maintain.

I recommend we create a `loopback-acl` module that exports this middleware. We can still land the `ACL` changes as is.

File: test/route-acl.test.js [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+      permission: 'ALLOW'
+    },
+    {
+      principalType: 'ROLE',
+      principalId: 'cs',
+      scopes: ['cancelOrder'],
+      permission: 'ALLOW'
+    }
+  ],
+  roles: {
```
Comment (ritch) [inline test/route-acl.test.js]: Why do we have to include roles here? Is this an advantage somehow?

Comment (raymondfeng) [inline test/route-acl.test.js]: Yes, this allows us to define role mappings without DB.

File: test/route-acl.test.js (hunk 2 above)
Comment (raymondfeng) [inline test/route-acl.test.js]: Can you point me to an example? Please note that my test doesn't boot up an app.

File: test/route-acl.test.js [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+      },
+      query: {
+        filter: {
+          limit: 100
+        }
+      }
+    };
+    var res = {};
+    handler(req, res, function(err) {
+      if (err) return done();
```
Comment (ritch) [inline test/route-acl.test.js]: this should assert the `res` has a status code or something similar... `err` could be included for many unrelated reasons / unrelated errors

File: test/route-acl.test.js [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        'Accept': 'application/json'
+      },
+      accessToken: {
+        userId: 'john',
+        appId: 'demo-app'
+      }
+    };
+    var res = {};
+    handler(req, res, function(err) {
+      if (err) return done();
```
Comment (ritch) [inline test/route-acl.test.js]: same as comment above... assert something about the `res`... `err` isn't enough

**Comment by ritch** [comment]
I'm closing this since it should not be merged. Instead we should create a new module that exports this middleware.

Also the tests should not use `loopback.*` classes directly.

**Comment by raymondfeng** [comment]
See https://github.com/strongloop-internal/scrum-loopback/issues/423

**Pull request closed without merging**